### ISSUE TITLE
Support Fetch API Request and Response

### ISSUE THREAD

**Issue by MatthewWid** [MatthewWid/better-sse#79]
Update the Session to be able to take in a [`Request`](https://developer.mozilla.org/en-US/docs/Web/API/Request) object from the [Fetch API](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API) and return in some way a [`Response`](https://developer.mozilla.org/en-US/docs/Web/API/Response) object.

Web streams and the Fetch API appear to be the predominant standard with JavaScript moving forward, and as such this will enable compatibility with the many libraries (#70, #78, #83) that expose `Request`/`Response` interfaces rather than the Node [`IncomingMessage`](https://nodejs.org/api/http.html#class-httpincomingmessage) and [`ServerResponse`](https://nodejs.org/api/http.html#class-httpserverresponse) objects.

---

The new API would simply add another overload to the session constructor, so you could now do one of:

```typescript
createSession(IncomingMessage, ServerResponse, options)
createSession(Http2ServerRequest, Http2ServerResponse, options)
createSession(Request, Response, options)
```

Which would be used like the following:

```typescript
import { createSession } from "better-sse";

const session = await createSession(request);

return session.getResponse();
```

If the session detects that it is passed a `Request` object it will defer calling the internal `initialize` method - which flushes the response head and sends preamble data - until the `getResponse` method has been called.

When it is finally invoked, a new `Response` object with a [`ReadableStream`](https://developer.mozilla.org/en-US/docs/Web/API/ReadableStream) as its body will be returned, given the parameters from the session constructor [`options`](https://matthewwid.github.io/better-sse/reference/api/#new-sessionstate--defaultsessionstatereq-incomingmessage--http2serverrequest-res-serverresponse--http2serverresponse-options-) argument as well as any additional parameters given to the method itself which are passed through to the [`Response` constructor](https://developer.mozilla.org/en-US/docs/Web/API/Response/Response), and the library will begin streaming data through it.

This may involve a refactor to where, even the if the `IncomingMessage`/`ServerResponse` objects are given, they are converted to be `Request`/`Response` objects internally so that a single, standardized interface can be used across different frameworks and runtimes rather than repeatedly [running down conditional trees with slow and janky `instanceof` checks](https://github.com/MatthewWid/better-sse/blob/c8c02ec49f76ab16b7ec092901afe27236711307/src/Session.ts#L210).

---

This may obsolete #77 as instead of making adapters implement a now-abstract `Session` class you would instead write an adapter to fit your runtime/frameworks lifecycle into the standard `Request`/`Response` objects, requiring no breaking changes to the current package API and overall being a much cleaner implementation.

---

Some possible reference implementations:

* [SvelteKit](https://github.com/sveltejs/kit/blob/896723519b1ed8595160b8ebaf1d56df0a3c1065/packages/kit/src/exports/node/index.js#L108)
* [`node-fetch-server`](https://github.com/mjackson/remix-the-web/blob/bb0ced8c1f6cb20caac34d719a9d84cffe0648b7/packages/node-fetch-server/src/lib/request-listener.ts#L126)

**Comment by go4cas** [comment]
@MatthewWid ... first off, thanks for the great library!

Question: once this issue has been implemented, will better-sse work in a Hono NodeJS app?

**Comment by MatthewWid** [comment]
@go4cas yes, that would be the case.

In *theory* you should even be able to use Hono with non-Node runtimes such as Deno and Bun, but I need to do more research and testing on that.

---

I don't have a fully formed idea of what the API would and _could_ look like at the moment, but something resembling this is the general workflow:

```typescript
app.get("/sse", async (c) => {
  const session = await createSession(c.req);

  // register to channels and/or push events

  return session.getResponse();
});
```

There might be some strangeness, however, with the order of things being returned whereby data is pushed to the session before the response is returned and headers are flushed. In which case, there needs to be some mechanism where we can access the session only after the response has been returned. For example:

```typescript
const session = await createSession(c.req); // flushing headers is deferred until `getResponse` is called

session.once("connected", () => {
  // safe to send data
});

return session.getResponse();
```

or,

```typescript
return (await createSession(c.req)).getResponse((session) => {
  // safe to send data
});
```

I'll have to give it more thought when I can find the time.

**Comment by TimSusa** [comment]
scratching behind my head, what you mean with that question, see my code screenshot:
<img width="589" alt="Screenshot 2024-12-11 at 17 53 09" src="https://github.com/user-attachments/assets/7a333ecb-ce81-4a95-9ea8-095762279a46" />
look!
<img width="894" alt="Screenshot 2024-12-11 at 17 51 03" src="https://github.com/user-attachments/assets/cca69aad-340d-4f8f-ac01-fb0d0f4a52f8" />

**Comment by MatthewWid** [comment]
A note about a tradeoff made in the design of this feature in the interest of performance:

I had originally intended to update the [`Session#push`](https://matthewwid.github.io/better-sse/reference/api/#sessionpush-data-unknown-eventname-string-eventid-string--this) method to go from a synchronous method that returned `this` to an asynchronous method returning a [`Promise`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise) that would resolve once the underlying data chunk with the event buffer data had flushed successfully to the kernel buffer.

This would allow users to `await` each event push so they could be assured the previous event was entirely flushed out of the server before the next event was pushed, as well as enabling error handling with try/catch in the unlikely case that the Node internals somehow throw from the call to [`res.write`](https://nodejs.org/api/http.html#responsewritechunk-encoding-callback). This also lines up nicely with the Promise returned from the [`WritableStreamDefaultWriter#write`](https://developer.mozilla.org/en-US/docs/Web/API/WritableStreamDefaultWriter/write) method for users taking advantage of the new support for the [Fetch API](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API).

When running the benchmark suite on the initial implementation, however, there was a **massive** performance dip when pushing events sequentially to a single session:

<details>
<summary>Benchmark results before (sync) and after (async)</summary>

Before (v0.14.1):

```
> getting-started@0.1.0 start
> node -r ts-node/register --max-old-space-size=3059 $INIT_CWD/index.ts

Benchmarking "Push events with sessions".
better-sse    x 743,530 ops/sec ±6.29% (77 runs sampled)
sse-channel   x 633,543 ops/sec ±34.30% (61 runs sampled)
sse           x  84,090 ops/sec ±67.30% (16 runs sampled)
```

After (v0.15.0-pre-alpha):

```
> getting-started@0.1.0 start
> node -r ts-node/register --max-old-space-size=3059 $INIT_CWD/index.ts

Benchmarking "Push events with sessions".
better-sse    x 254,836 ops/sec ±7.67% (29 runs sampled)
sse-channel   x 543,135 ops/sec ±38.95% (55 runs sampled)
sse           x  66,610 ops/sec ±69.04% (18 runs sampled)
```
</details>

I then profiled each of the individual function calls made from within the `push` method (and even wrote my own shoddy performance monitor, [PerfWtf](https://gist.github.com/MatthewWid/cb857414f5e03126dd8e9f8b754bd7e6), to do so):

<details>
<summary>Internal call time results before (sync) and after (async)</summary>

Below, we run the `push` method 100,000 times and measure the time it takes for each internal function call to run. `avg` is the average time across all calls, `total` is the total time of all calls added together and `count` is the number of times called.

Before (v0.14.1):

```
[
  {
    "name": "push.emit",
    "last": 488.868353,
    "avg": 0.00013091960080397172,
    "total": 13.09209099999805,
    "count": 100001
  },
  {
    "name": "connection.sendChunk",
    "last": 488.868127,
    "avg": 0.0009020020799584587,
    "total": 90.20201200000619,
    "count": 100002
  },
  {
    "name": "buffer.push",
    "last": 488.867336,
    "avg": 0.000939518284817146,
    "total": 93.9527680000001,
    "count": 100001
  },
  {
    "name": "session.flush",
    "last": 488.868193,
    "avg": 0.001035708655826989,
    "total": 103.57293700001168,
    "count": 100002
  },
  {
    "name": "session.push",
    "last": 488.868419,
    "avg": 0.00234575096249035,
    "total": 234.57744200000275,
    "count": 100001
  }
]
```

After (v0.15.0-pre-alpha):

```
[
  {
    "name": "push.emit",
    "last": 528.992031,
    "avg": 0.0001698360916391072,
    "total": 16.983779000002443,
    "count": 100001
  },
  {
    "name": "buffer.push",
    "last": 528.989922,
    "avg": 0.0002482124278757111,
    "total": 24.821490999999384,
    "count": 100001
  },
  {
    "name": "connection.sendChunk",
    "last": 528.991739,
    "avg": 0.0013535996680066982,
    "total": 135.362674000006,
    "count": 100002
  },
  {
    "name": "session.flush",
    "last": 528.991796,
    "avg": 0.0015048935621287708,
    "total": 150.49236600000157,
    "count": 100002
  },
  {
    "name": "session.push",
    "last": 528.992087,
    "avg": 0.002342907740922412,
    "total": 234.2931169999815,
    "count": 100001
  }
]
```
</details>

As you can see, however, each of the internal calls appears to run in about the same time in both the old and new versions.

As such, I thought that it must be the overhead of creating the promise _itself_ that is causing the slowdown, not any changes to how we are encoding or writing to the network. And indeed, when turning the `push` method back into a synchronous function but changing nothing else about its internal logic we see the benchmarks return back to their expected numbers:

<details>
<summary>Benchmark results before (sync) and after (sync with fetch support)</summary>

Before (v0.14.1):

```
> getting-started@0.1.0 start
> node -r ts-node/register --max-old-space-size=3059 $INIT_CWD/index.ts

Benchmarking "Push events with sessions".
better-sse x 608,262 ops/sec ±5.16% (81 runs sampled)
sse-channel x 667,447 ops/sec ±24.65% (65 runs sampled)
sse x 202,873 ops/sec ±61.21% (21 runs sampled)
```

After (v0.15.0-pre-alpha):

```
> getting-started@0.1.0 start
> node -r ts-node/register --max-old-space-size=3059 $INIT_CWD/index.ts

Benchmarking "Push events with sessions".
better-sse x 564,249 ops/sec ±6.26% (76 runs sampled)
sse-channel x 629,567 ops/sec ±38.98% (64 runs sampled)
sse x 174,238 ops/sec ±54.96% (23 runs sampled)
```
</details>

While I expected slightly lowered numbers, I did not expect a staggering ~60% performance drop simply from wrapping two function calls in a promise.

To me, this is unacceptable, so I have made the decision to retain the existing public API for now and keep `push` as a synchronous method.

This means that, just like before, in the unlikely scenario where an error is thrown from the Node internals when attempting to write to the network (most often these originate from some kernel-level error), wrapping calls to `Session#push` in a try/catch will **not** catch the error, even if you are using the Fetch API:

```typescript
const session = await createSession(request)

// Will not work
try {
    session.push("I'm going to to crash your network driver >:)")
} catch (error) {
    // Never reached for asynchronous throws
}
```

You might consider adding a global [exception handler](https://nodejs.org/api/process.html#event-uncaughtexception) if providing the Node [HTTP API](https://nodejs.org/api/http.html) or a global [rejection handler](https://nodejs.org/api/process.html#event-unhandledrejection) if providing the [Fetch API](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API).

There have been no complaints about this behavior so far, so I believe this is a worthy trade-off to maintain performance.

---

Works continues on the [`fetch-request-response` branch](https://github.com/MatthewWid/better-sse/tree/fetch-request-response).

**Comment by MatthewWid** [comment]
Implemented in v0.15.0.

See the [original pull request](https://github.com/MatthewWid/better-sse/pull/91) and [release announcement](https://github.com/MatthewWid/better-sse/releases/tag/v0.15.0) for more details.
